File: go/mechanisms/evm/exact/client/eip2612.go
package client

import (
	"context"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"

	"github.com/coinbase/x402/go/extensions/eip2612gassponsor"
	"github.com/coinbase/x402/go/mechanisms/evm"
)

// SignEip2612Permit signs an EIP-2612 permit authorizing the Permit2 contract
// to spend tokens. This creates a gasless off-chain signature that the
// facilitator can submit on-chain via x402Permit2Proxy.settleWithPermit().
//
// The permittedAmount must match the Permit2 permitted.amount exactly, as the
// proxy contract enforces permit2612.value == permittedAmount.
func SignEip2612Permit(
	ctx context.Context,
	signer evm.ClientEvmSignerWithReadContract,
	tokenAddress string,
	tokenName string,
	tokenVersion string,
	chainID *big.Int,
	deadline string,
	permittedAmount string,
) (*eip2612gassponsor.Info, error) {
	owner := signer.Address()
	spender := evm.PERMIT2Address
	normalizedToken := evm.NormalizeAddress(tokenAddress)

	// Query the current EIP-2612 nonce from the token contract
	nonceResult, err := signer.ReadContract(
		ctx,
		normalizedToken,
		evm.EIP2612NoncesABI,
		"nonces",
		common.HexToAddress(owner),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to read EIP-2612 nonce: %w", err)
	}

	nonce, ok := nonceResult.(*big.Int)
	if !ok {
		return nil, fmt.Errorf("unexpected nonce type: %T", nonceResult)
	}

	// Parse deadline
	deadlineBig, ok := new(big.Int).SetString(deadline, 10)
	if !ok {
		return nil, fmt.Errorf("invalid deadline: %s", deadline)
	}

	// Construct EIP-712 domain for the token's permit function
	domain := evm.TypedDataDomain{
		Name:              tokenName,
		Version:           tokenVersion,
		ChainID:           chainID,
		VerifyingContract: normalizedToken,
	}

	types := evm.GetEIP2612EIP712Types()

	approvalAmount, ok := new(big.Int).SetString(permittedAmount, 10)
	if !ok {
		return nil, fmt.Errorf("invalid permitted amount: %s", permittedAmount)
	}

	message := map[string]interface{}{
		"owner":    owner,
		"spender":  spender,
		"value":    approvalAmount,
		"nonce":    nonce,
		"deadline": deadlineBig,
	}

	// Sign the EIP-2612 permit
	signatureBytes, err := signer.SignTypedData(ctx, domain, types, "Permit", message)
	if err != nil {
		return nil, fmt.Errorf("failed to sign EIP-2612 permit: %w", err)
	}

	return &eip2612gassponsor.Info{
		From:      owner,
		Asset:     normalizedToken,
		Spender:   spender,
		Amount:    approvalAmount.String(),
		Nonce:     nonce.String(),
		Deadline:  deadline,
		Signature: evm.BytesToHex(signatureBytes),
		Version:   "1",
	}, nil
}

File: typescript/packages/mechanisms/evm/src/exact/client/eip2612.ts
import { getAddress } from "viem";
import { eip2612PermitTypes, eip2612NoncesAbi, PERMIT2_ADDRESS } from "../../constants";
import { ClientEvmSigner } from "../../signer";
import type { Eip2612GasSponsoringInfo } from "../extensions";

export type Eip2612PermitSigner = Pick<ClientEvmSigner, "address" | "signTypedData"> & {
  readContract: NonNullable<ClientEvmSigner["readContract"]>;
};

/**
 * Signs an EIP-2612 permit authorizing the Permit2 contract to spend tokens.
 *
 * This creates a gasless off-chain signature that the facilitator can submit
 * on-chain via `x402Permit2Proxy.settleWithPermit()`.
 *
 * The `permittedAmount` must match the Permit2 `permitted.amount` exactly, as the
 * proxy contract enforces `permit2612.value == permittedAmount`.
 *
 * @param signer - The client EVM signer (must support readContract for nonce query)
 * @param tokenAddress - The ERC-20 token contract address
 * @param tokenName - The token name (from paymentRequirements.extra.name)
 * @param tokenVersion - The token version (from paymentRequirements.extra.version)
 * @param chainId - The chain ID
 * @param deadline - The deadline for the permit (unix timestamp as string)
 * @param permittedAmount - The Permit2 permitted amount (must match exactly)
 * @returns The EIP-2612 gas sponsoring info object
 */
export async function signEip2612Permit(
  signer: Eip2612PermitSigner,
  tokenAddress: `0x${string}`,
  tokenName: string,
  tokenVersion: string,
  chainId: number,
  deadline: string,
  permittedAmount: string,
): Promise<Eip2612GasSponsoringInfo> {
  const owner = signer.address;
  const spender = getAddress(PERMIT2_ADDRESS);

  // Query the current EIP-2612 nonce from the token contract
  const nonce = (await signer.readContract({
    address: tokenAddress,
    abi: eip2612NoncesAbi,
    functionName: "nonces",
    args: [owner],
  })) as bigint;

  // Construct EIP-712 domain for the token's permit function
  const domain = {
    name: tokenName,
    version: tokenVersion,
    chainId,
    verifyingContract: tokenAddress,
  };

  const approvalAmount = BigInt(permittedAmount);

  const message = {
    owner,
    spender,
    value: approvalAmount,
    nonce,
    deadline: BigInt(deadline),
  };

  // Sign the EIP-2612 permit
  const signature = await signer.signTypedData({
    domain,
    types: eip2612PermitTypes,
    primaryType: "Permit",
    message,
  });

  return {
    from: owner,
    asset: tokenAddress,
    spender,
    amount: approvalAmount.toString(),
    nonce: nonce.toString(),
    deadline,
    signature,
    version: "1",
  };
}
